cibuildwheel/logger.py

```diff
@@ -0,0 +1,192 @@
+import os
+import time
+import sys
+from typing import Optional, Union
+
+DEFAULT_FOLD_PATTERN = ('{name}', '')
+FOLD_PATTERNS = {
+    'azure': ('##[group]{name}', '##[endgroup]'),
+    'travis': ('travis_fold:start:{name}', 'travis_fold:end:{name}'),
+    'github': ('::group::{name}', '::endgroup::{name}'),
+}
+
+PLATFORM_IDENTIFIER_DESCIPTIONS = {
+    'manylinux_x86_64': 'Manylinux x86_64',
+    'manylinux_i686': 'Manylinux i686',
+    'manylinux_aarch64': 'Manylinux aarch64',
+    'manylinux_ppc64le': 'Manylinux ppc64le',
+    'manylinux_s390x': 'Manylinux s390x',
+    'win32': 'Windows 32bit',
+    'win_amd64': 'Windows 64bit',
+    'macosx_x86_64': 'macOS x86_64',
+}
+
+
+class Logger:
+    fold_mode: str
+    colors_enabled: bool
+    active_build_identifier: Optional[str] = None
+    build_start_time: Optional[float] = None
+    step_start_time: Optional[float] = None
+    active_fold_group_name: Optional[str] = None
+
+    def __init__(self):
+        if 'AZURE_HTTP_USER_AGENT' in os.environ:
+            self.fold_mode = 'azure'
+            self.colors_enabled = True
+
+        elif 'GITHUB_ACTIONS' in os.environ:
+            self.fold_mode = 'github'
+            self.colors_enabled = True
+
+        elif 'TRAVIS' in os.environ:
+            self.fold_mode = 'travis'
+            self.colors_enabled = True
+
+        elif 'APPVEYOR' in os.environ:
+            self.fold_mode = 'disabled'
+            self.colors_enabled = True
+
+        else:
+            self.fold_mode = 'disabled'
+            self.colors_enabled = file_supports_color(sys.stdout)
+
+    def build_start(self, identifier: str):
+        c = self.colors
+        description = build_description_from_identifier(identifier)
+        print()
+        print(f'{c.bold}{c.blue}Building {identifier} wheel{c.end}')
+        print(f'{description}')
+        print()
+
+        self.build_start_time = time.time()
+        self.active_build_identifier = identifier
+
+    def build_end(self):
+        assert self.build_start_time is not None
+        assert self.active_build_identifier is not None
+        self.step_end()
+
+        c = self.colors
+        duration = time.time() - self.build_start_time
+
+        print()
+        print(f'{c.green}✓ {c.end}{self.active_build_identifier} finished in {duration:.2f}s')
+        self.build_start_time = None
+        self.active_build_identifier = None
+
+    def step(self, step_description: str):
+        self.step_end()
+        self.step_start_time = time.time()
+        self._start_fold_group(step_description)
+
+    def step_end(self, success=True):
+        if self.step_start_time is not None:
+            self._end_fold_group()
+            c = self.colors
+            duration = time.time() - self.step_start_time
+            if success:
+                print(f'{c.green}✓ {c.end}{duration:.2f}s'.rjust(78))
```

Comment:
Open question: do we always want to print times (ánd not fold those?)

Reply:
Hmm. Well I liked having some confirmation that the step completed (the green tick) - it's reassuring. The time was kinda added opportunistically but could be very helpful for debugging slow builds (or preventing people from trying to optimise the wrong bit!). 

But I'll try rolling all that inside the fold group to see how it looks.